pkg/audit/stats_reporter.go

```diff
@@ -2,111 +2,142 @@ package audit
+	violationsM       metric.Int64ObservableGauge
+	auditDurationM    metric.Float64Histogram
+	lastRunStartTimeM metric.Float64ObservableGauge
+	lastRunEndTimeM   metric.Float64ObservableGauge
+	meter             metric.Meter
 )
 
 func init() {
-	if err := register(); err != nil {
+	var err error
+	meter = otel.GetMeterProvider().Meter("gatekeeper")
+
+	violationsM, err = meter.Int64ObservableGauge(
+		violationsMetricName,
+		metric.WithDescription("Total number of audited violations"),
+	)
+
+	if err != nil {
 		panic(err)
 	}
-}
 
-func register() error {
-	views := []*view.View{
-		{
-			Name:        violationsMetricName,
-			Measure:     violationsM,
-			Aggregation: view.LastValue(),
-			TagKeys:     []tag.Key{enforcementActionKey},
-		},
-		{
-			Name:        auditDurationMetricName,
-			Measure:     auditDurationM,
-			Aggregation: view.Distribution(1*60, 3*60, 5*60, 10*60, 15*60, 20*60, 40*60, 80*60, 160*60, 320*60),
-		},
-		{
-			Name:        lastRunStartTimeMetricName,
-			Measure:     lastRunStartTimeM,
-			Aggregation: view.LastValue(),
-		},
-		{
-			Name:        lastRunEndTimeMetricName,
-			Measure:     lastRunEndTimeM,
-			Aggregation: view.LastValue(),
-		},
+	auditDurationM, err = meter.Float64Histogram(
+		auditDurationMetricName,
+		metric.WithDescription("Latency of audit operation in seconds"))
+	if err != nil {
+		panic(err)
 	}
-	return view.Register(views...)
-}
 
-func (r *reporter) reportTotalViolations(enforcementAction util.EnforcementAction, v int64) error {
-	ctx, err := tag.New(
-		context.Background(),
-		tag.Insert(enforcementActionKey, string(enforcementAction)))
+	lastRunStartTimeM, err = meter.Float64ObservableGauge(
+		lastRunStartTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run starting time"),
+	)
 	if err != nil {
-		return err
+		panic(err)
+	}
+
+	lastRunEndTimeM, err = meter.Float64ObservableGauge(
+		lastRunEndTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run ending time"),
+	)
+	if err != nil {
+		panic(err)
 	}
 
-	return r.report(ctx, violationsM.M(v))
+	view.Register(sdkmetric.NewView(
+		sdkmetric.Instrument{Name: auditDurationMetricName},
+		sdkmetric.Stream{
+			Aggregation: sdkmetric.AggregationExplicitBucketHistogram{
+				Boundaries: []float64{1 * 60, 3 * 60, 5 * 60, 10 * 60, 15 * 60, 20 * 60, 40 * 60, 80 * 60, 160 * 60, 320 * 60},
+			},
+		},
+	))
 }
 
-func (r *reporter) reportLatency(d time.Duration) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
+func (r *reporter) registerCallback() error {
+	_, err1 := meter.RegisterCallback(r.observeTotalViolations, violationsM)
+	_, err2 := meter.RegisterCallback(r.observeRunEnd, lastRunEndTimeM)
+	_, err3 := meter.RegisterCallback(r.observeRunStart, lastRunStartTimeM)
+	return errors.Join(err1, err2, err3)
+}
+
+func (r *reporter) observeTotalViolations(_ context.Context, o metric.Observer) error {
+	for k, v := range r.totalViolationsPerEnforcementAction {
```

Comment:
This (and all other observe functions) need some sort of thread-safety mechanism (locks, atomic variables, etc.) in this case, a read mutex (since you're using a write mutex below).